pageserver/src/page_service.rs

```diff
@@ -304,8 +309,21 @@ fn page_service_conn_main(
         .context("could not set TCP_NODELAY")?;
 
     let mut conn_handler = PageServerHandler::new(conf, auth);
-    let pgbackend = PostgresBackend::new(socket, auth_type, None, true)?;
-    pgbackend.run(&mut conn_handler)
+    let pgbackend =
+        PostgresBackend::new(socket, auth_type, None, true).map_err(anyhow::Error::from)?;
+    pgbackend.run(&mut conn_handler).map_err(|err| {
+        let root_cause_io_err_kind = err
+            .root_cause()
+            .downcast_ref::<io::Error>()
+            .map(|e| e.kind());
+
+        if let Some(io::ErrorKind::ConnectionReset) = root_cause_io_err_kind {
+            error::PageServerError::PostgresIO(PostgresIOError::ConnectionReset)
+        } else {
+            error::PageServerError::Internal(err)
+        }
+    })?;
```

Comment:
can be simplified by removing the question mark and `Ok(())`. Because map err transforms the error to correct type already